### DISCUSSION TITLE
Failed to read the request form. Request body too large. The max request body size is 30000000 bytes

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/aspnetcore#48708] [Q&A] [answered]
We're stuck on making .NET accept big file uploads.

We develop on Linux, deploy inside containers, and run behind Nginx reverse proxy.

We develop using `dotnet build` and `dotnet bin/Project.dll`, and we deploy using `dotnet publish -o /path` and run inside a docker container using `dotnet Project.dll`.

In the development environment we do not have a `web.config` file, but when we use the `publish` command, it creates a default `web.config` file for us.

We have already added these lines to our `Startup.cs` file:

```
services.Configure<FormOptions>(x =>
{
    x.ValueLengthLimit = int.MaxValue;
    x.MultipartBodyLengthLimit = int.MaxValue;
    x.MultipartHeadersLengthLimit = int.MaxValue;
});
```

But it seems that this is not enough. In our development environment, we do not have a `web.config` file so we can't configure our API using that. We tried to configure `Kestrel` using this code:

```
services.Configure<Microsoft.AspNetCore.Server.Kestrel.KestrelServerOptions>(options =>
{
    options.Limits.MaxRequestBodySize = int.MaxValue;
});
```

But we got this error:

> error CS0234: The type or namespace name 'KestrelServerOptions' does not exist in the namespace 'Microsoft.AspNetCore.Server.Kestrel' (are you missing an assembly reference?)

We can use per action configuration as follow, and it works.

```
[RequestFormLimits(ValueLengthLimit = int.MaxValue,
        MultipartBodyLengthLimit = long.MaxValue)]
[DisableRequestSizeLimit]
```

But we want to enable large file upload for **all** of our APIs.

How should we do it?

**Comment by david-acker** [comment, marked as answer]
Could you try configuring `Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions` instead? 

API Docs: [KestrelServerOptions](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserveroptions?view=aspnetcore-7.0)

It looks like `KestrelServerOptions` actually used be in the `Microsoft.AspNetCore.Server.Kestrel` namespace back in [ASP.NET Core 1.1](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.kestrelserveroptions?view=aspnetcore-1.1), but has since been moved to `Microsoft.AspNetCore.Server.Kestrel.Core`. That old version of the docs is the first search result I'm getting for "KestrelServerOptions" on Google, which is a bit confusing.

**Reply by Nefcanto** [reply]
@david-acker thank you. That solved it. I didn't check the version of `KestrelServerOptions`.

**Comment by davidfowl** [comment]
All of the information with code samples are on this page: https://learn.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-7.0

Here's how you tweak the limit in Kestrel:

https://learn.microsoft.com/en-us/aspnet/core/mvc/models/file-uploads?view=aspnetcore-7.0#kestrel-maximum-request-body-size

It's not showing the WebApplicationBuilder pattern:

```C#
builder.WebHost.ConfigureKestrel(options =>
{
    options.Limits.MaxRequestBodySize = A big number; // Maybe avoid int.MaxValue
});
```
